server/protocols/jmap/src/test/java/org/apache/james/jmap/model/JmapMDNTest.java

```diff
@@ -122,4 +128,50 @@ public void textBodyIsCompulsory() {
             .isInstanceOf(IllegalStateException.class);
     }
 
+    @Test
+    public void generateMDNMessageShouldUseFromHeaderWhenNoSenderHeader() throws Exception {
+        String senderAddress = "sender@local";
+        Message originMessage = Message.Builder.of()
+            .setMessageId("45554@local.com")
+            .setFrom(senderAddress)
+            .setBody("body", StandardCharsets.UTF_8)
+            .build();
+
+        assertThat(
+            MDN.generateMDNMessage(originMessage,
+                new MockMailboxSession("user@localhost.com"))
+                .getTo())
+            .extracting(address -> (Mailbox) address)
+            .extracting(Mailbox::getAddress)
+            .containsExactly(senderAddress);
+    }
+
+    @Test
+    public void messageWithoutMessageIdHeaderShouldBeInvalid() throws Exception {
+        Message originMessage = Message.Builder.of()
+            .setFrom("sender@local")
+            .setBody("body", StandardCharsets.UTF_8)
+            .build();
+
+        assertThatThrownBy(() ->
+            MDN.generateMDNMessage(originMessage,
+                new MockMailboxSession("user@localhost.com"))
+                .getTo())
+            .isInstanceOf(InvalidOriginMessageForMDNNotificationException.class);
+    }
+
+    @Test
+    public void messageWithoutSenderAndFromHeaderShouldBeInvalid() throws Exception {
+        Message originMessage = Message.Builder.of()
+            .setMessageId("45554@local.com")
+            .setBody("body", StandardCharsets.UTF_8)
+            .build();
+
+        assertThatThrownBy(() ->
+            MDN.generateMDNMessage(originMessage,
+                new MockMailboxSession("user@localhost.com"))
+                .getTo())
+            .isInstanceOf(InvalidOriginMessageForMDNNotificationException.class);
+    }
+
```

Comment:
you miss some tests: sender + nofrom, sender + from, multiple sender (?), multiple from

Reply:
No API for multi senders, I won't test that case here, if you don't mind...